src/io/rsm/rsm_client.hpp

```diff
@@ -131,6 +153,117 @@ class RsmClient {
 
     return TimedOut{};
   }
+
+  /// AsyncRead methods
+  void SendAsyncReadRequest(ReadRequestT req) {
+    MG_ASSERT(!async_read_);
+
+    ReadRequest<ReadRequestT> read_req;
+    read_req.operation = req;
```

Comment:
A `ReadRequestT` gets copied once this function is called and once here. This is expensive for large data structures (and all our `Messages` contain a vector, a map or some other dynamically allocated data structure).

Reply:
There were actually 3 copies, since we also cache the previous request.
We still need at least one deep copy because of the cached request IMO. But I will take it by const& and move where possible.
Same goes for `WriteRequestT`